### PR TITLE
Cytoband search

### LINKED ISSUE #1906: Add variant search by cytoband
When using filter (SvFilters) for Clinical Structural Variants in Scout there is no option to use AND instead of OR. This hampers any attempt at refined combinatorial searches. Also, it is not possible to search for chromosomal bands as only chromosome number 1, 2, 3 etc are accepted. Could for example a radio button indicating AND respectively OR be added to filter options in Scout? All the best Tord Jonson (Lund)

### PR DESCRIPTION
**Pull request by northwestwitch**
fix #1906 

Adds support for cytoband search in SNV and SV variants filter.

The cytoband files for build 37 and 38 are taken from UCSC: http://genome.ucsc.edu/cgi-bin/hgTables?db=hg38&hgta_group=map&hgta_track=cytoBand&hgta_table=cytoBand&hgta_doSchema=describe+table+schema. From these files I've removed the lines for non-canonical chromosomes.

Note that these cytoband coordinardinates are 0 based. Shall we make them 1-based instead? 

The filter looks like this in both SNV and SVs filter
image

The cytoband start and end selects are dynamic and their content depends on the selected chromosome (I've turned the chromosome field into a select, and it's much better I think!)
Once a cytoband start is selected it updates the start coordinate in the filters form. End coordinate works in the same way

**How to test**:
1. Cytobands collection can be populated in 2 ways:
 - `scout setup demo` or `scout setup`
 - `scout load cytobands` (--build to specify which build, otherwise both builds are used) 
2. After populating the cytobands collection in a local instance of Scout, go to the demo case, variants page. Use the cytoband start and end selects to filter variants
3.  Go to the structural variants page. Use the cytoband start and end selects to filter variants.

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [x] tests executed by CR

### CODE REVIEW

**Comment by northwestwitch** [comment]
@moonso there is a conceptual error in the codefactorsettings: It's checking for code that is not used right away, when the page is loaded, but when the user interacts with the variants filters form. 
These functions are indeed used (I can even introduce tests) but only on form submissions.

image

Can we skip checking the javascript in codefactor??

File: scout/server/blueprints/variants/forms.py [hunk 1]
```diff
@@ -98,7 +99,9 @@ class FiltersForm(VariantFiltersForm):
     clinsig_confident_always_returned = BooleanField("CLINSIG Confident")
     spidex_human = SelectMultipleField("SPIDEX", choices=SPIDEX_CHOICES)
 
-    chrom = TextField("Chromosome", [validators.Optional()])
+    chrom = SelectField("Chromosome", [validators.Optional()], choices=("",) + CHROMOSOMES)
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/forms.py:102]: Turned this field into a select with pre-determined values

File: scout/server/blueprints/variants/forms.py [hunk 2]
```diff
@@ -138,7 +141,9 @@ class SvFiltersForm(VariantFiltersForm):
     clingen_ngi = IntegerField("ClinGen NGI obs")
     swegen = IntegerField("SweGen obs")
 
-    chrom = TextField("Chromosome", [validators.Optional()])
+    chrom = SelectField("Chromosome", [validators.Optional()], choices=("",) + CHROMOSOMES)
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/forms.py:144]: Same here!

**Comment by northwestwitch** [comment]
This is done and ready for review. I could really use some help to understand the reason why it's not passing the "Run tests for different versions / Tests (3.6, 3.6) (push) " test

**Comment by moonso** [comment]
I will have a look

**Comment by northwestwitch** [comment]
> I will have a look

Thanks! Could be some method missing in previous version of python / mongo, but I'm not sure

**Comment by moonso** [comment]
Did you try to run the tests locally @northwestwitch ? I did and get the same failure so you probably need to check that

**Comment by northwestwitch** [comment]
Of course I tried and it works perfectly! 🙄 What python and mongodb
versions are you using?

**Comment by moonso** [comment]
Oh it seems like I didn't have mongo running 🙈 . Anyway I'm in a very stressful situation preparing for tomorrow so I will not have time to dig into this today. You can go into the workflow and try to figure out what went wrong

**Comment by northwestwitch** [comment]
> Oh it seems like I didn't have mongo running 🙈 . Anyway I'm in a very stressful situation preparing for tomorrow so I will not have time to dig into this today. You can go into the workflow and try to figure out what went wrong

Ok thanks for nothing then! 😆

**Comment by dnil** [comment]
FWIW it does not run on my local either (py3.6.8, mongo 3.4.10), but that may just be due to current debugging.

I'm missing one piece of logic though: cytobands are like stepwise hierarchies of more detailed coordinates, right? So if I say 1p3 I mean the entire 1p31-1p36, 1p36 includes 1p36.11-1p36.21 and 1p36.1 includes 1p36.11-1p36.13.

**Comment by northwestwitch** [comment]
> FWIW it does not run on my local either (py3.6.8, mongo 3.4.10), but that may just be due to current debugging.

You don't happen to see any error message that could point to the error? The log of the test is not clear, I just see deprecation warnings.

> I'm missing one piece of logic though: cytobands are like stepwise hierarchies of more detailed coordinates, right? So if I say 1p3 I mean the entire 1p31-1p36, 1p36 includes 1p36.11-1p36.21 and 1p36.1 includes 1p36.11-1p36.13.

I didn't want to make it complicated and especially I didn't want to leave much freedom to the user when comes to insert text for searches, because freedom in general comes with bugs. But you are right. What if I include just the bands located at the lowest level then? Do you have any suggestion?

**Comment by dnil** [comment]
Oh, and yes, add 1 to the displayed coordinates. BED file coordinates are an odd interval type 0-based, start inclusive, end non-inclusive. VCF is 1 based, inclusive.

**Comment by dnil** [comment]
My reflex would have been to write (or import) a parser for general cytoband syntax and have only one field for chr bands etc. But you are painfully right this is more error prone than a few dropdowns. I may be able to live with that, but perhaps we should ask one of the cyto-persons if they cringe too much to use it? 😜 

A value error from WTF if I recall correctly. I'll switch back in just a moment and check.

**Comment by dnil** [comment]
```
2020-05-25 10:19:44 k1315C02VR0WRHTDH scout.adapter.mongo.variant[11460] DEBUG Fetching variants from internal_id
2020-05-25 10:19:44 k1315C02VR0WRHTDH scout.adapter.mongo.query[11460] DEBUG Building a mongo query for internal_id
2020-05-25 10:19:44 k1315C02VR0WRHTDH scout.adapter.mongo.query[11460] DEBUG Querying category sv
2020-05-25 10:19:44 k1315C02VR0WRHTDH scout.adapter.mongo.query[11460] DEBUG Set variant type to clinical
2020-05-25 10:19:44 k1315C02VR0WRHTDH scout.adapter.mongo.query[11460] DEBUG Adding panel and genes-related parameters to the query
2020-05-25 10:19:44 k1315C02VR0WRHTDH scout.adapter.mongo.query[11460] INFO mongo query: {'case_id': 'internal_id', 'category': 'sv', 'variant_type': 'clinical', 'panels': {'$in': ['panel1']}, '$and': [{'$or': [{'chromosome': 'None'}, {'end_chrom': 'None'}]}]}
2020-05-25 10:19:44 k1315C02VR0WRHTDH flask.app[11460] ERROR Exception on /cust000/643594/sv/variants [GET]
Traceback (most recent call last):
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/flask/app.py", line 2292, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/flask/app.py", line 1815, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/flask/app.py", line 1718, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/flask/_compat.py", line 35, in reraise
    raise value
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/dannil/sandbox/scout/scout/server/utils.py", line 28, in decorated_function
    return render_template(template_name, **context)
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/flask/templating.py", line 135, in render_template
    context, ctx.app)
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/flask/templating.py", line 117, in _render
    rv = template.render(context)
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "/Users/dannil/sandbox/scout/scout/server/blueprints/variants/templates/variants/sv-variants.html", line 146, in top-level template code
    <button name="page" type="submit" class="btn btn-outline-secondary mx-auto d-block" value={{ page + 1 }}>Next page</button>
  File "/Users/dannil/sandbox/scout/scout/server/templates/layout_bs4.html", line 1, in top-level template code
    {% extends "bootstrap4.html" %}
  File "/Users/dannil/sandbox/scout/scout/server/templates/bootstrap4.html", line 18, in top-level template code
    {% block body %}
  File "/Users/dannil/sandbox/scout/scout/server/templates/bootstrap4.html", line 21, in block "body"
    {% block content %}
  File "/Users/dannil/sandbox/scout/scout/server/templates/layout_bs4.html", line 81, in block "content"
    {% block content_main %}{% endblock %}
  File "/Users/dannil/sandbox/scout/scout/server/blueprints/variants/templates/variants/sv-variants.html", line 49, in block "content_main"
    <div class="card-body">{{ sv_filters(form, institute, case) }}</div>
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/jinja2/runtime.py", line 579, in _invoke
    rv = self._func(*arguments)
  File "/Users/dannil/sandbox/scout/scout/server/blueprints/variants/templates/variants/utils.html", line 344, in template
    {{ form.chrom(class="form-control") }}
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/wtforms/fields/core.py", line 155, in __call__
    return self.meta.render_field(self, kwargs)
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/wtforms/meta.py", line 56, in render_field
    return field.widget(field, **render_kw)
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/wtforms/widgets/core.py", line 323, in __call__
    for val, label, selected in field.iter_choices():
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/wtforms/fields/core.py", line 454, in iter_choices
    for value, label in self.choices:
ValueError: not enough values to unpack (expected 2, got 0)
2020-05-25 10:19:44 k1315C02VR0WRHTDH werkzeug[11460] INFO 127.0.0.1 - - [25/May/2020 10:19:44] "GET /cust000/643594/sv/variants?variant_type=clinical&gene_panels=panel1 HTTP/1.1" 500 -
```

**Comment by dnil** [comment]
Looks like the start values for page / query are messing up?

**Comment by northwestwitch** [comment]
> have only one field for chr bands etc.

Using the dropdowns you could still do the same search, given that the cytobands in the dropdown are ordered by ascending coordinates. And it's not error prone. 
If for instance you want to filter for the region 1p31-1p36 you select the first option in the cytoband start with 1p36 and the last one in the cytoband end with 1p31. They would get used to it I think

**Comment by dnil** [comment]
Agreed, the three dropdowns will allow all the options of the one text field. With fewer manual errors. It all hinges on if they can endure the near physical pain of not being able to explain to the student that that is not how cytobands work. 😆

**Comment by northwestwitch** [comment]
My latest commit should fix the error you posted before @dnil, could you check again? It's still not passing the tests/commands/test_setup_cmd.py::test_setup_database test

**Comment by dnil** [comment]
Hm, no? Still same thing. I can see if I can dig a little after lunch.

I noticed there is still use of the old CYTOBANDS loaded "constant" in parse/variant/coordinates.py. Wouldn't it be nice to replace it, making it build aware?
Or did you want to put that in another PR for #1908?

**Comment by northwestwitch** [comment]
> Or did you want to put that in another PR for #1908?

I think I'm already having problems as it is, let's leave it for another one! 🙏

File: scout/server/blueprints/variants/forms.py [hunk 3]
```diff
@@ -138,10 +142,12 @@ class SvFiltersForm(VariantFiltersForm):
...
-    chrom = TextField("Chromosome", [validators.Optional()])
+    chrom = SelectField("Chromosome", [validators.Optional()], choices=CHROMOSOME_OPTIONS)
+    cytoband_start = SelectField("Cytoband start")
+    cytoband_end = SelectField("Cytoband end")
     start = IntegerField("Start position", [validators.Optional()])
     end = IntegerField("End position", [validators.Optional()])
 
-    filter_variants = SubmitField(label="Filter variants")
-    clinical_filter = SubmitField(label="Clinical filter")
+    filter_variants = SubmitField(label="Filter variants", choices=[])
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/forms.py:151]: These ones are not selects, they do not need choices, right?

Comment (dnil) [reply scout/server/blueprints/variants/forms.py:151]: Just cut and paste misses.

**Comment by dnil** [comment]
I'm still seeing local test issues for "clinical": 
```
tests/server/blueprints/variants/test_variants_views.py::test_variants_clinical_filter FAILED                                                                                                         [ 85%]

================================================================================================= FAILURES ==================================================================================================
_______________________________________________________________________________________ test_variants_clinical_filter _______________________________________________________________________________________

app = <Flask 'scout.server.app'>
institute_obj = {'internal_id': 'cust000', '_id': 'cust000', 'display_name': 'test_institute', 'sanger_recipients': ['john@doe.com', '...889), 'updated_at': datetime.datetime(2020, 5, 25, 17, 17, 36, 54879), 'coverage_cutoff': 10, 'frequency_cutoff': 0.01}
case_obj = {'_id': 'internal_id', 'analysis_date': datetime.datetime(2016, 10, 12, 14, 0, 46), 'assignee': None, 'assignees': [], ...}

    def test_variants_clinical_filter(app, institute_obj, case_obj):
    ...
         # containing the variant above
>           assert updated_var["_id"] in str(resp.data)
E           assert 'e287358e4abec884c39a4ed4363cc130' in 'b\'<!DOCTYPE html>\\n<head>\\n\\t<meta charset="utf-8">\\n\\t<meta name="viewport" content="width=device-width, initi...ilters_form\\\'];\\n        the_form.submit();\\n      };\\n    });\\n  </script>\\n\\n\\n\\t\\n\\n</body>\\n</html>\''
E            +  where 'b\'<!DOCTYPE html>\\n<head>\\n\\t<meta charset="utf-8">\\n\\t<meta name="viewport" content="width=device-width, initi...ilters_form\\\'];\\n        the_form.submit();\\n      };\\n    });\\n  </script>\\n\\n\\n\\t\\n\\n</body>\\n</html>\'' = str(b'<!DOCTYPE html>\n<head>\n\t<meta charset="utf-8">\n\t<meta name="viewport" content="width=device-width, initial-scal...ument.forms[\'filters_form\'];\n        the_form.submit();\n      };\n    });\n  </script>\n\n\n\t\n\n</body>\n</html>')
E            +    where b'<!DOCTYPE html>\n<head>\n\t<meta charset="utf-8">\n\t<meta name="viewport" content="width=device-width, initial-scal...ument.forms[\'filters_form\'];\n        the_form.submit();\n      };\n    });\n  </script>\n\n\n\t\n\n</body>\n</html>' = <<class 'pytest_flask.plugin.JSONResponse'> 21964 bytes [200 OK]>.data

tests/server/blueprints/variants/test_variants_views.py:60: AssertionError
------------------------------------------------------------------------------------------- Captured stdout setup -------------------------------------------------------------------------------------------
```

File: scout/load/cytoband.py [hunk 4]
```diff
@@ -1,4 +1,7 @@
 import logging
+import os.path
```
Comment (dnil) [inline scout/load/cytoband.py:2]: :eyes:

**Comment by northwestwitch** [comment]
Passes all tests except the codefactor on javascript

image

**Comment by northwestwitch** [comment]
😨  @dnil 

image

**Comment by dnil** [comment]
That did work locally.. 😆

File: scout/load/cytoband.py [hunk 5]
```diff
@@ -23,23 +24,24 @@ def load_cytobands(resource, build, adapter):
...
-            cytoband_obj = dict(
-                _id=generate_md5_key([build, chrom, band]),
-                band=band,
-                chrom=str(chrom),  # 3
-                start=str(int(fields[1]) + 1),  # 58600000
-                stop=str(int(fields[2]) + 1),  # 63800000
-                build=str(build),  # "37" or "38"
-            )
-            cytobands.append(cytoband_obj)
+    lines = get_file_handle(resource)
```
Comment (dnil) [inline scout/load/cytoband.py:27]: :+1:

Comment (dnil) [reply scout/load/cytoband.py:27]: Thanks - I saw it afterwards.. Reverting to unzipped would be totally fine with me as well.

Comment (dnil) [reply scout/load/cytoband.py:27]: But, this is nice and consistent, and shows some promise to work also without --editable.

**Comment by dnil** [comment]
For reference, from chat:

Ahhhh… Right! Its the MANIFEST file that specifies what ends up in the install package. This is what we have now:
```
# static files
recursive-include scout/resources *.gz
recursive-include scout/demo *
```
Which explains why txt files are fine in demo, but only gz in resources. Let's try without -e tomorrow, but agreed - not now!

**Comment by dnil** [comment]
Core functionality is also good on local:
Screenshot 2020-05-25 at 21 12 32

Screenshot 2020-05-25 at 21 13 31

:+1:

**Comment by northwestwitch** [comment]
@dnil this is ready for the "final final" review now!

File: scout/commands/load/cytoband.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+from scout.load.cytoband import load_cytobands
+from scout.resources import cytoband_files
+from scout.server.extensions import store
+
+
+@click.command("cytobands", short_help="Load cytobands")
+@click.option(
+    "--build",
+    type=click.Choice(["37", "38"]),
+    help="What genome build should be used. If no choice update 37 and 38.",
```
Comment (dnil) [inline scout/commands/load/cytoband.py:13]: should be loaded perhaps?

File: scout/commands/load/cytoband.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    """Populate cytobands collection
+
+    Args:
+        build(str): "37" or "38"
+    """
+    # Remove previous cytoband objects from cytoband collection
+    store.cytoband_collection.drop()
+
+    if build is None:
+        builds = cytoband_files.keys()
```
Comment (dnil) [inline scout/commands/load/cytoband.py:26]: :+1:

File: scout/commands/load/cytoband.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    else:
+        builds = [build]
+    # Look cytobands for each chromosome build
+    for genome_build in builds:
+        resource_path = cytoband_files.get(genome_build)
+        load_cytobands(cytoband_files.get(genome_build), genome_build, store)
```
Comment (dnil) [inline scout/commands/load/cytoband.py:32]: You probably meant to use the resource_path here.. 😉

File: scout/load/cytoband.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    LOG.debug(f"Reading cytoband file for genome build {build}")
+
+    if path.exists(resource) is False:
+        LOG.error(f"Resource {resource} could not be found.")
+        return
+
+    lines = get_file_handle(resource)
+
+    for line in lines:
+        # Line will look like this:
```
Comment (dnil) [inline scout/load/cytoband.py:30]: What do you think about ignoring header \#-starting lines? Not a must - we don't have it in the current ones.

File: scout/load/cytoband.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+        # 3	58600000	63800000	p14.2	gneg
+        fields = line.split("\t")
+        chrom = fields[0]
```
Comment (dnil) [inline scout/load/cytoband.py:33]: And how about ignoring "chr" here? Not an issue with current files, but does add an implicit constraint on the files.

File: scout/load/setup.py [hunk 11]
```diff
@@ -119,6 +120,10 @@ def setup_scout(
     else:
         exac_lines = fetch_exac_constraint()
 
+    # Load cytobands into cytoband collection
+    for genome_build, cytobands_path, in cytoband_files.items():
```
Comment (dnil) [inline scout/load/setup.py:124]: I find trailing commas scary. Just saying. 😆

File: tests/commands/test_setup_cmd.py [hunk 12]
```diff
@@ -23,7 +22,8 @@ def test_setup_database(mock_app):
     assert runner
 
     # test the CLI command for seting up scout
-    result = runner.invoke(cli, ["setup", "demo"], input="y")
+    result = runner.invoke(cli, ["setup", "demo"])
```
Comment (dnil) [inline tests/commands/test_setup_cmd.py:25]: Looks ok to me, but perhaps it was part of debug?

**Comment by northwestwitch** [comment]
> Yay! Only tiny change suggestions and a couple of very optional file reading things.

Thanks for the review, a couple of things were really helpful to prevent potential bugs. Aways good to have another pair of eyes to check your code!

Regarding the code in the test:
`    result = runner.invoke(cli, ["setup", "demo"], input="y")`

Modified to:
`    result = runner.invoke(cli, ["setup", "demo"])`

The test was wrong in the first place, because the setup demo command doesn't ask you to confirm that you want to drop a previous database, it's the `scout setup database` (non-demo) that requires the user to confirm.

File: scout/load/setup.py (hunk 11 above)
Comment (northwestwitch) [reply scout/load/setup.py]: Me too! This was a typo, good that you noticed!

File: scout/load/cytoband.py (hunk 10 above)
Comment (northwestwitch) [reply scout/load/cytoband.py]: Done

File: scout/load/cytoband.py (hunk 9 above)
Comment (northwestwitch) [reply scout/load/cytoband.py:30]: Yes

**Pull request merged**